protocols/autonat/src/protocol.rs

```diff
@@ -0,0 +1,310 @@
+// Copyright 2018 Parity Technologies (UK) Ltd.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use crate::structs_proto;
+use async_trait::async_trait;
+use futures::io::{AsyncRead, AsyncWrite, AsyncWriteExt};
+use libp2p_core::{upgrade, Multiaddr, PeerId};
+use libp2p_request_response::{ProtocolName, RequestResponseCodec};
+use prost::Message;
+use std::{convert::TryFrom, io};
+
+#[derive(Clone, Debug)]
+pub struct AutoNatProtocol;
+
+impl ProtocolName for AutoNatProtocol {
+    fn protocol_name(&self) -> &[u8] {
+        b"/libp2p/autonat/1.0.0"
+    }
+}
+
+#[derive(Clone)]
+pub struct AutoNatCodec;
+
+#[async_trait]
+impl RequestResponseCodec for AutoNatCodec {
+    type Protocol = AutoNatProtocol;
+    type Request = DialRequest;
+    type Response = DialResponse;
+
+    async fn read_request<T>(
+        &mut self,
+        _: &AutoNatProtocol,
+        io: &mut T,
+    ) -> io::Result<Self::Request>
+    where
+        T: AsyncRead + Send + Unpin,
+    {
+        let bytes = upgrade::read_length_prefixed(io, 1024)
+            .await
+            .map_err(|e| io::Error::new(io::ErrorKind::InvalidData, e))?;
```

Comment:
I don't think these should need `.map_err` anymore after https://github.com/libp2p/rust-libp2p/pull/2111.